# Web-scraping: сбор данных из баз данных и интернет-источников

*Алла Тамбовцева*

## Практикум 1. Введение в парсинг с BeautifulSoup: поиск по тэгам

Импортируем необходимые библиотеки. Нам понадобится модуль `requests` для отправки запросов, для «подключения» к странице и получения её содержимого в виде строки, и функция `BeautifulSoup` из библиотеки `bs4` для удобного поиска по полученной строке:

In [1]:
import requests

In [2]:
from bs4 import BeautifulSoup

Подключаемся к странице сайта с материалами [math-info.hse.ru](http://math-info.hse.ru):

In [3]:
page = requests.get("http://math-info.hse.ru/2024-25")

In [4]:
page

<Response [404]>

Объект, который мы сохранили в `page`, имеет особый тип *request*, он же запрос. Из него можно извлечь исходный код страницы в виде обычного текста (тип *string*). Однако в нашем случае мы этого сделать не сможем, потому что в ответ на запрос мы получили результат с кодом 404. Известная ошибка, которая возникает тогда, когда ссылка ведёт вникуда. Исправляемся – добавляем слэш в конце ссылки, в данном случае это важно:

In [5]:
page = requests.get("http://math-info.hse.ru/2024-25/")
page

<Response [200]>

Теперь код 200, это означает, что запрос был благополучно отправлен и на него был получен ответ. Извлечём из полученного ответа код HTML в виде обычной строки:

In [6]:
# page.text

Выполнять поиск по такой строке с тэгами не очень удобно (даже если вы знакомы с регулярными выражениями), поэтому преобразуем строку в объект типа `BeautifulSoup`. Такой объект внешне несильно отличается от обычной строки, однако внутри по структуре похож на словарь, и это сходство значительно упрощает поиск по тэгам и атрибутам.

In [7]:
soup = BeautifulSoup(page.text)

In [8]:
# внешне то же самое, что и ранее
# soup

На объектах `BeautifulSoup` определены методы `.find()` и `.find_all()`. Оба метода возвращают фрагменты кода HTML, которые соответствуют критериям поиска, различие заключается в том, что метод `.find()` предназначен для поиска одного совпадения (если критериям поиска соответствует несколько элементов на странице, то берётся только первый), а метод `.find_all()` – для поиска всех совпадений. В первом случае результат возвращается в виде одного элемента типа `BeautifulSoup`, а во втором – в виде списка таких элементов. Давайте попробуем что-то поискать!

Найдём сначала заголовок первого уровня – он здесь точно один (тэг `<h1>`):

In [9]:
soup.find("h1")

<h1 class="firstHeading" id="firstHeading" lang="ru">Заглавная страница</h1>

Результат – фрагмент кода HTML со всеми тэгами. Как получить «чистый» текст? Запросить текст с помощью атрибута `.text`:

In [10]:
soup.find("h1").text

'Заглавная страница'

Теперь найдём все заголовки второго уровня:

In [11]:
soup.find_all("h2")

[<h2>Содержание</h2>,
 <h2><span id="Школа_лингвистики"></span><span class="mw-headline" id=".D0.A8.D0.BA.D0.BE.D0.BB.D0.B0_.D0.BB.D0.B8.D0.BD.D0.B3.D0.B2.D0.B8.D1.81.D1.82.D0.B8.D0.BA.D0.B8">Школа лингвистики</span></h2>,
 <h2><span id="ОП_Политология"></span><span class="mw-headline" id=".D0.9E.D0.9F_.D0.9F.D0.BE.D0.BB.D0.B8.D1.82.D0.BE.D0.BB.D0.BE.D0.B3.D0.B8.D1.8F">ОП Политология</span></h2>,
 <h2><span id="Факультет_креативных_индустрий"></span><span class="mw-headline" id=".D0.A4.D0.B0.D0.BA.D1.83.D0.BB.D1.8C.D1.82.D0.B5.D1.82_.D0.BA.D1.80.D0.B5.D0.B0.D1.82.D0.B8.D0.B2.D0.BD.D1.8B.D1.85_.D0.B8.D0.BD.D0.B4.D1.83.D1.81.D1.82.D1.80.D0.B8.D0.B9">Факультет креативных индустрий</span></h2>,
 <h2><span id="ОП_Вычислительные_социальные_науки"></span><span class="mw-headline" id=".D0.9E.D0.9F_.D0.92.D1.8B.D1.87.D0.B8.D1.81.D0.BB.D0.B8.D1.82.D0.B5.D0.BB.D1.8C.D0.BD.D1.8B.D0.B5_.D1.81.D0.BE.D1.86.D0.B8.D0.B0.D0.BB.D1.8C.D0.BD.D1.8B.D0.B5_.D0.BD.D0.B0.D1.83.D0.BA.D0.B8">ОП Вычислительные соц

Сохраним полученный с названием `h2` и поработаем с ним (и не только)!

In [12]:
h2 = soup.find_all("h2")

### Задача 1

Сохраните в новый список `hh` только названия образовательных программ или факультетов/школ, то есть все элементы списка `h2`, кроме первого и последнего.

In [13]:
hh = h2[1:-1]
print(hh)

[<h2><span id="Школа_лингвистики"></span><span class="mw-headline" id=".D0.A8.D0.BA.D0.BE.D0.BB.D0.B0_.D0.BB.D0.B8.D0.BD.D0.B3.D0.B2.D0.B8.D1.81.D1.82.D0.B8.D0.BA.D0.B8">Школа лингвистики</span></h2>, <h2><span id="ОП_Политология"></span><span class="mw-headline" id=".D0.9E.D0.9F_.D0.9F.D0.BE.D0.BB.D0.B8.D1.82.D0.BE.D0.BB.D0.BE.D0.B3.D0.B8.D1.8F">ОП Политология</span></h2>, <h2><span id="Факультет_креативных_индустрий"></span><span class="mw-headline" id=".D0.A4.D0.B0.D0.BA.D1.83.D0.BB.D1.8C.D1.82.D0.B5.D1.82_.D0.BA.D1.80.D0.B5.D0.B0.D1.82.D0.B8.D0.B2.D0.BD.D1.8B.D1.85_.D0.B8.D0.BD.D0.B4.D1.83.D1.81.D1.82.D1.80.D0.B8.D0.B9">Факультет креативных индустрий</span></h2>, <h2><span id="ОП_Вычислительные_социальные_науки"></span><span class="mw-headline" id=".D0.9E.D0.9F_.D0.92.D1.8B.D1.87.D0.B8.D1.81.D0.BB.D0.B8.D1.82.D0.B5.D0.BB.D1.8C.D0.BD.D1.8B.D0.B5_.D1.81.D0.BE.D1.86.D0.B8.D0.B0.D0.BB.D1.8C.D0.BD.D1.8B.D0.B5_.D0.BD.D0.B0.D1.83.D0.BA.D0.B8">ОП Вычислительные социальные науки</span></h2>

### Задача 2

Извлеките из каждого элемента списка `hh` текст и сохраните полученные названия программ/факультетов/школ в список `headers`.

In [14]:
# с циклом
headers = []
for h in hh:
    headers.append(h.text)
print(headers)

['Школа лингвистики', 'ОП Политология', 'Факультет креативных индустрий', 'ОП Вычислительные социальные науки', 'ОП Психология']


In [15]:
# со списковым включением
headers = [h.text for h in hh]
print(headers)

['Школа лингвистики', 'ОП Политология', 'Факультет креативных индустрий', 'ОП Вычислительные социальные науки', 'ОП Психология']


### Задача 3

Найдите на странице список курсов ОП «Политология» (их просто больше всего) и сохраните его в переменную `polit`.

In [16]:
# находим все ненумерованные списки и берем третий
polit = soup.find_all("ul")[2]

### Задача 4

Используя `polit`, получите список названий курсов и назовите его `names`.

In [17]:
# элементы списка с тэгами
names_raw = polit.find_all("li")

# извлекаем текст из каждого элемента
names = [li.text for li in names_raw]
print(names)

['Высшая математика (1 курс)', 'НИС Введение в регрессионный анализ (2 курс)', 'Программирование для всех (основы работы с Python) (1 курс магистратуры)', 'Основы анализа данных в Python (3 курс)', 'НИС Основы анализа данных в Python (3 курс)', 'Введение в теорию вероятностей и математическую статистику (1 курс)']


### Задача 5

Извлеките из каждого элемента `names` названия предметов «в чистом» виде, без указаний на то, на каком курсе они читаются, и сохраните результаты в новый список `names_clean`.

Подсказка: метод `.split()` для строк.

In [18]:
# на примере одного элемента: разбиваем строку
# по ( и из списка выбираем первый элемент (до скобки)

n = names[0]
n.split(" (")[0]

'Высшая математика'

In [19]:
# теперь для всех
names_clean = [n.split(" (")[0] for n in names]
print(names_clean)

['Высшая математика', 'НИС Введение в регрессионный анализ', 'Программирование для всех', 'Основы анализа данных в Python', 'НИС Основы анализа данных в Python', 'Введение в теорию вероятностей и математическую статистику']


### Задача 6

Дан перечень id для курсов, созданный с помощью функции `range()`:

In [20]:
ids = range(1, 7)
print(*ids)

1 2 3 4 5 6


Используя `ids` и список `names_clean`, создайте словарь `info`, где ключами являются id, а значениями – названия курсов.

In [21]:
# склеиваем элементы списков попарно
print(*zip(ids, names_clean))

(1, 'Высшая математика') (2, 'НИС Введение в регрессионный анализ') (3, 'Программирование для всех') (4, 'Основы анализа данных в Python') (5, 'НИС Основы анализа данных в Python') (6, 'Введение в теорию вероятностей и математическую статистику')


In [22]:
# склеиваем элементы списков попарно
# и превращаем список пар в словарь

info = dict(zip(ids, names_clean))
print(info)

{1: 'Высшая математика', 2: 'НИС Введение в регрессионный анализ', 3: 'Программирование для всех', 4: 'Основы анализа данных в Python', 5: 'НИС Основы анализа данных в Python', 6: 'Введение в теорию вероятностей и математическую статистику'}


**Дополнение.** Конечно, чтобы сохранить информацию в более серьезном виде, результаты стоит выгрузить в файл. В данном случае информации мало, но для примера посмотрим, как простой список сохранить в табличном виде в формате Excel (для длинных списков точно будет актуально, чтобы не запускать каждый раз при работе код для их получения с нуля).

Импортируем библиотеку `pandas` с сокращенным названием:

In [23]:
import pandas as pd

Превратим список с названиями курсов в таблицу – датафрейм pandas:

In [25]:
pd.DataFrame(names_clean)

0
0                                  Высшая математика
1                НИС Введение в регрессионный анализ
2                          Программирование для всех
3                     Основы анализа данных в Python
4                 НИС Основы анализа данных в Python
5  Введение в теорию вероятностей и математическу...

Можем сохранить и выгрузить ее в файл:

In [26]:
df = pd.DataFrame(names_clean)
df.to_excel("names.xlsx")

Теперь файл `names.xlsx` лежит в рабочей папке Jupyter (рядом с текущим ipynb-файлом). Если по перечню файлов в *Home* непонятно, что это за папка, можно импортировать модуль `os` для работы с операционной системой и выяснить путь к этой папке:

In [27]:
import os

# get current working directory
os.getcwd()

'/Users/allat/Dropbox/ВШЭ - работа/Кафедра высшей математики/2024-2025/Web-scraping'